### PR TITLE
Preparations before introducing checked NPEs

### PR DESCRIPTION
**Pull request by sjrd**
Including the fix of #4755.

### CODE REVIEW

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 1]
```diff
@@ -2349,11 +2349,19 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
...
-
+            if (!tarray.tpe.isNullable) {
+              val array = finishTransformExpr(tarray)
+              val index = finishTransformExpr(tindex)
+              val select = ArraySelect(array, index)(elemType)
+              contTree(select)
+            } else {
+              withNewTempLocalDefs(targs) { (localDefs, cont1) =>
+                val List(arrayDef, indexDef) = localDefs
+                val select = ArraySelect(arrayDef.newReplacement, indexDef.newReplacement)(elemType)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:2361]: The semantics of this node are very counter-intuitive, but specified as such:
https://lampwww.epfl.ch/~doeraene/sjsir-semantics/#sec-sjsir-arrayget-runtime-semantics-evaluation

I assume the reasons for it being specified that way lay somewhere at the intersection of how the relevant JVM and JS constructs are specified.

In any case, I think this warrants a comment, *why* the shorter form is valid if `tarray` is not-nullable.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 2]
```diff
@@ -2362,12 +2370,23 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
...
-            val array = finishTransformExpr(tarray)
-            val index = finishTransformExpr(tindex)
             val elemType = cursoryArrayElemType(arrayTpe)
-            val select = ArraySelect(array, index)(elemType)
-            val tunboxedValue = foldAsInstanceOf(tvalue, elemType)
-            contTree(Assign(select, finishTransformExpr(tunboxedValue)))
+            if (!tarray.tpe.isNullable) {
+              val array = finishTransformExpr(tarray)
+              val index = finishTransformExpr(tindex)
+              val select = ArraySelect(array, index)(elemType)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:2377]: Same here, RE `ArraySelect` having somewhat counter-intuitive semantics.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala [hunk 3]
```diff
@@ -195,4 +209,35 @@ class ArrayTest {
...
+    @noinline def testNoInline[T](array: Array[T], value: T): Unit = {
+      testGeneric(array, value)
+    }
+
+    @inline def test[T](array: Array[T], value: T): Unit = {
+      testNoInline(array, value)
+      testGeneric(array, value)
+    }
+
+    val nullArrayRef = nullOf[Array[AnyRef]]
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala:235]: Nit: Inline this val? (also below).

**Review by gzm0 — APPROVED** [review]
Only some nits.

A note aside: I vaguely remember having seen a comment about evaluation order for NPEs not being 100% JVM compliant for method application on the NPE changes. However, I cannot find it anymore. Did this change? If no, what are our decision criteria for when we want to guarantee evaluation order w.r.t. NPEs and when not.

**Comment by gzm0** [comment]
Also, there is a typo in the main commit:

>  we must make sure to evaluation all
three arguments

to **evaluate **

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala (hunk 1 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:2373]: > I assume the reasons for it being specified that way lay somewhere at the intersection of how the relevant JVM and JS constructs are specified.

Yes. In JS, `array[index]` throws a `TypeError` before evaluating `index` when `array === null`. We have something similar for method calls.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala (hunk 3 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala]: The `val`s are actually necessary for the test not to be moot. I added a comment about that.

**Review by gzm0 — APPROVED** [review]

**Comment by gzm0** [comment]
> In this case, it's not about being consistent with the JVM as much as being consistent with our own non-optimized semantics. Our optimizer must not introduce a UB NPE in a program that does not expose one.

Of course. That makes perfect sense.

> If we wanted to delay NPEs until after the arguments are evaluated, like the JVM does, we would have to generate assignments of all receivers and arguments to all method calls in the entire program before doing the calls. Even with Unchecked semantics. That would explode code size. So we chose to specify NPEs as coming earlier instead.

Hm... I realize this is totally unrelated to this PR, but wouldn't it be OK to change the throwing point in unchecked mode? UB can do far worse things than changing where something fails, can't it?

**Comment by sjrd** [comment]
> Hm... I realize this is totally unrelated to this PR, but wouldn't it be OK to change the throwing point in unchecked mode? UB can do far worse things than changing where something fails, can't it?

No, it's emphatically not OK, not in that direction. From checked to unchecked, you can *delay* UB or remove it or do something else *from the point that UB happened onwards*. But you cannot have UB happen *earlier* than the check that would be performed in checked mode. Otherwise, even when your tests pass in checked mode, you have no guarantee about the correctness of the unchecked mode.

**Pull request merged**